[design] Functions for explicitly fetched signature algorithms

levitte opened the pull request:
This design goes into more details what was outlined in the design for
[fetching composite (PKEY) algorithms and using them].

It also changes what functionality will be used for this.  The design for
signature was originally to add modified initializers for DigestSign and
DigestVerify, but recent OTC discussions redirected us to have a closer look
at `EVP_PKEY_sign()` and `EVP_PKEY_verify()`.

Finally, it also takes into account the need to specify the signature
to be verified against with `EVP_PKEY_verify()` streaming functions,
which has been discussed in #22357.

Related to #22357 (in progress), #22129 (merged), and openssl/project#231

[fetching composite (PKEY) algorithms and using them]:
    https://github.com/openssl/openssl/blob/07c769d8267ae6f9066d76e403accc0b94082d6e/doc/designs/fetching-composite-algorithms.md

levitte committed:
[design] Functions for explicitly fetched signature algorithms

This design goes into more details what was outlined in the design for
[fetching composite (PKEY) algorithms and using them].

It also changes what functionality will be used for this.  The design for
signature was originally to add modified initializers for DigestSign and
DigestVerify, but recent OTC discussions redirected us to have a closer look
at EVP_PKEY_sign() and EVP_PKEY_verify().

[fetching composite (PKEY) algorithms and using them]:
    ./fetching-composite-algorithms.md

levitte commented:
This design is still missing some provider interfaces to be added.  I'm currently looking more closely at that.

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,125 @@
+int EVP_PKEY_verify_final(EVP_PKEY_CTX *ctx);
+
+int EVP_PKEY_verify_recover_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_verify_recover_final(EVP_PKEY_CTX *ctx);
```
This one needs the `unsigned char *rout, size_t *routlen` parameters.

levitte replied:
Fixed

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,125 @@
+
+There are two ways this could be implemented:
+
+1.  through an added provider function that would work like keymgmt's
+    `query_operation_name` function, but would return a key type name
+    instead:
+
+    ``` C
+    # define OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE         26
+    OSSL_CORE_MAKE_FUNC(const char *, signature_query_key_type, (void))
+    ```
+
+2.  through a gettable `OSSL_PARAM`, using the param identity "keytype"
````
This returns just one type but presumably RSA-PSS-WITH-SHA256 could use both RSA and RSA-PSS keys.

levitte replied:
Eh gods, that.  How I regret making RSA-PSS a separate key type...

levitte replied:
I'll have a think about that

jamuir replied:
fyi -- I have been working on at adding RSA-OAEP as a new key type in #22344

levitte replied:
I don't like it, but I'll reduce this part to have only the first option, and let it return `const char **`.

t8m reviewed (commented):
This looks good so far, although I need to think through how applications should change to be able to use these new functions but still work with old providers and legacy.

levitte commented:
> This looks good so far, although I need to think through how applications should change to be able to use these new functions but still work with old providers and legacy.

This new functionality won't support legacy.  This is about *explicitly fetched* algorithms, there's nothing legacy about that.

The streaming functions (update, final) will also not be possible to use with old providers.  The one-shot functions should continue to work transparently, however.

slontis commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,160 @@
+# define OSSL_FUNC_SIGNATURE_SIGN_UPDATE            26
+# define OSSL_FUNC_SIGNATURE_SIGN_FINAL             27
+OSSL_CORE_MAKE_FUNC(int, signature_sign_update, (void *ctx,  unsigned char *sig,
+                                                 size_t *siglen, size_t sigsize,
```
Should output params be last?

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 61 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

levitte commented:
Rebased...

Also, I had forgotten that this was still draft, it should have been made ready for review a while ago.

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 61 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR has been closed.  It was waiting for the creator to make requested changes but it has not been updated for 90 days.

t8m commented:
@levitte please drop the unrelated fuzz-corpora submodule change.

levitte commented:
I squashed and rebased too, while I was at it

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,155 @@
+Public API - Add variants of `EVP_PKEY_CTX` functionality
+---------------------------------------------------------
+
+Through OTC discussions, it's been determined that the most suitable APIs to
+touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
+`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
+associated functions.  They can be extended to accept an explicitly fetched
+algorithm of the right type, and to be able to process infinite amounts of
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
```
As @bernd-edlinger and @davidben wrote in this https://github.com/openssl/openssl/pull/23240#issuecomment-1916968985 and following comments this is potentially confusing to users and we might want to separate the API calls for _signing a message digest_ from calls for _signing a message_.

On the other hand the EVP_DigestSign set of API calls are misdesigned and ugly so my proposal is to:

1. Add an alias EVP_PKEY_sign_digest() to the existing EVP_PKEY_sign() call, and EVP_PKEY_verify_digest() to the existing EVP_PKEY_verify() call.
2. Add functions EVP_PKEY_sign_message(), EVP_PKEY_verify_message() that would be implemented by these ED25519, RSA-SHA256 or similar algorithms.

levitte replied:
Depending on how you read it, *message* is quite confusing too.  If you read PKCS#1 carefully, you'll note that RSASP1 and RSAVP1 also take a (limited size) message, *represented* as an integer 0 → n - 1 (they talk about "message representative", which I read to mean that the integer is a representation of the message, a limited size octet string that was passed to O2ISP, or that will get passed through I2OSP).

levitte replied:
Thinking a bit about it, it could be that *primitive* and *data* provide a better distinction, with the semantics that *primitive* only allow a one-shot style of operation, while *data* allows a stream style of operation.  So we'd end up with:

- `EVP_PKEY_sign_primitive_init()` and `EVP_PKEY_sign_primitive()` (with the old `EVP_PKEY_sign_init()` and `EVP_PKEY_sign()` as aliases)
- `EVP_PKEY_sign_data_init()`, `EVP_PKEY_sign_data_update()` and `EVP_PKEY_sign_data_final()`

... and similar for any other stuff we want to give the same treatment.

t8m replied:
Yeah, that works for me.

t8m replied:
Now we need to think about the provider side interface.

t8m replied:
Could we just map the new APIs to the digest_sign_init()/digest_verify_init() provider functions with adding appropriate aliases through macros and add just new calls as needed?

levitte replied:
I've changed things around a bit, hopefully legibly

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,155 @@
+Added stream functionality:
+
+``` C
+int EVP_PKEY_sign_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,
````
These would be named EVP_PKEY_sign_data_update and EVP_PKEY_sign_data_final in my proposal.

levitte replied:
You have changed your mind along the way...  `message` → `data`

t8m replied:
Yeah, I was kind of undecided whether message or data is better - your choice. :grin:

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,155 @@
+Added stream functionality:
+
+``` C
+int EVP_PKEY_verify_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_verify_final(EVP_PKEY_CTX *ctx);
````
Similarly to above these would be verify_data_update and verify_data_final.

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,155 @@
+int EVP_PKEY_verify_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_verify_final(EVP_PKEY_CTX *ctx);
+
+int EVP_PKEY_verify_recover_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_verify_recover_final(EVP_PKEY_CTX *ctx,
+                                  unsigned char *rout, size_t *routlen);
```
I do not think the verify_recover should have a streaming variant.

levitte replied:
Hmm, ok

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,155 @@
+``` C
+int EVP_PKEY_verify_init_ex2(EVP_PKEY_CTX *pctx,
+                             EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
+int EVP_PKEY_verify_recover_init_ex2(EVP_PKEY_CTX *pctx,
+                                     EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
````
I am not so sure about usefulness of verify_recover for composite signature algorithms. IMO this is pretty much RSA specific thing.

simo5 commented:
FWIW you may want to ake in consideration this message from NIST:
https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/Mf2kemwwreY/m/KArjoIhxAQAJ

The part to take in consideration is the ability to have separate access layers to the functions, technically this could be completely a fips provider only thing, but it may be a good idea to provide this kind of api in general for testing with vectors. specfically for non deterministic signatures.

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

levitte commented:
Following @t8m's recent review comments, I did a bit of a rewrite.  Please review the changes

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

levitte commented on doc/designs/fetching-composite-algorithms.md:
```diff
@@ -43,68 +43,12 @@ int EVP_PKEY_derive_init_ex2(EVP_PKEY_CTX *ctx, EVP_KEYEXCH *exchange,
-explicit fetches of simple algorithms, say "RSA".  In that case, the caller
-will need to pass necessary auxiliary parameters through the `OSSL_PARAM` or
-a call to a corresponding `set_params` function.
+Detailed proposal for these APIs will are prepared in other design
```
Fixed

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,181 @@
+int EVP_PKEY_sign_primitive_init(EVP_PKEY_CTX *pctx,
+                                 EVP_SIGNATURE *algo,
+                                 const OSSL_PARAM params[]);
+int EVP_PKEY_sign_primitive(EVP_PKEY_CTX *ctx,
+                            unsigned char *sig, size_t *siglen, size_t sigsize,
+                            const unsigned char *tbs, size_t tbslen);
```
I assumed this will be just an alias to EVP_PKEY_sign().

levitte replied:
Implementation detail. But yeah

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,181 @@
+``` C
+int EVP_PKEY_sign_data_init(EVP_PKEY_CTX *pctx, EVP_SIGNATURE *algo,
+                            const OSSL_PARAM params[]);
+int EVP_PKEY_sign_data_update(EVP_PKEY_CTX *ctx,
````
IMO we should have also an EVP_PKEY_sign_data() one shot function with a respective provider call that could be possibly implemented as combined update/final call for providers that would not bother implementing the one shot call.

levitte replied:
Sure.  Would you think it's acceptable that the provider *only* implements the oneshot function (i.e. not update or final)?

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,181 @@
+int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
+                               unsigned char *sig, size_t siglen,
+                               size_t sigsize);
+/* Update and final */
+int EVP_PKEY_verify_data_update(EVP_PKEY_CTX *ctx,
+                                const unsigned char *in, size_t inlen);
+int EVP_PKEY_verify_data_final(EVP_PKEY_CTX *ctx);
```
Similarly to signature I'd propose having also EVP_PKEY_verify_data() oneshot call.

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,181 @@
+                                const unsigned char *in, size_t inlen);
+int EVP_PKEY_verify_data_final(EVP_PKEY_CTX *ctx);
+```
+
````
Nit: extra empty line.

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,181 @@
+OSSL_CORE_MAKE_FUNC(const char **, signature_query_key_type, (void))
+```
+
+Furthermore, the public API above requires added provider functionality:
+
+``` C
+# define OSSL_FUNC_SIGNATURE_SIGN_UPDATE            26
+# define OSSL_FUNC_SIGNATURE_SIGN_FINAL             27
+OSSL_CORE_MAKE_FUNC(int, signature_sign_update, (void *ctx,
+                                                 const unsigned char *in,
+                                                 size_t inlen))
+OSSL_CORE_MAKE_FUNC(int, signature_sign_final, (void *ctx,  unsigned char *sig,
+                                                size_t *siglen, size_t sigsize))
+
+# define OSSL_FUNC_SIGNATURE_VERIFY_UPDATE          28
+# define OSSL_FUNC_SIGNATURE_VERIFY_FINAL           29
+OSSL_CORE_MAKE_FUNC(int, signature_verify_update,
+                    (void *ctx, const unsigned char *in, size_t inlen))
+/*
+ * signature_verify_final requires that the signature to be verified against
+ * is specified via an OSSL_PARAM.
+ */
+OSSL_CORE_MAKE_FUNC(int, signature_verify_final, (void *ctx))
````
I assume that we need to reflect the EVP api calls in the providers interface as well?

levitte replied:
No, please no. As a provider author, I *protest* this proliferation of a gazillion functions that do the same essential thing

t8m replied:
I do not understand how we can implement the proposed EVP calls in terms of the existing provider calls.

levitte replied:
The only additional calls I can see making sense would be `signature_sign_data_init` and `signature_verify_data_init`.  If that was all you meant, then sure.

But for example, I don't see why we'd need to have two different oneshot functions.

t8m replied:
I was actually thinking about not adding additional _init functions because the existing ones could cover both sign_data and sign_primitive. But then you of course need two different oneshot functions because otherwise you cannot distinguish whether you want the primitive operation or not.

t8m replied:
If we have a different init operation for the primitive and data calls then the oneshot op can be the same, that's true.

levitte replied:
Actually...  I don't see much of a need to have different inits, even.  `primitive` vs `data` is really semantic sugar for the application, 'cause in the end, what differentiates the "primitive" vs "data" functionality is that the former requires calling `EVP_PKEY_CTX_set_signature_md()` while the latter doesn't (for "data", it should even be prohibited - not settable - since the "data" algorithms are expected to be composite algorithms, i.e. hash function implied).

In terms of oneshot, the difference between "primitive" and "data" is really that the former might error with some sort of error indicating that the input (tbs) is too large.

The flip side with having alternative function indexes (one for "primitive" and one for "data") is the possible implication that you should be able to have both.  That's a source of unnecessary error and confusion for provider authors.

levitte replied:
In the end, there are a number of possible indicators to see if an `EVP_SIGNATURE` can be considered "primitive", which can be checked in libcrypto:

- The presence or not of `OSSL_FUNC_SIGNATURE_SIGN_UPDATE` and `OSSL_FUNC_SIGNATURE_VERIFY_UPDATE`.
- The presence or not of `OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE`
- If there is a `OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE`, does its response match the key type of the given (through `EVP_PKEY_CTX`) `EVP_PKEY`?

t8m replied:
I am afraid this will be very confusing to provider authors.

levitte replied:
With my provider author hat on, I disagree.  I find the multitude of diverse functions to implement a much bigger pile of confusion...  and error prone outside of the provider's control (hard question: did libcrypto call the right variant of, say, update, for this use case?)

paulidale reviewed (dismissed):
Just a thought, but might it be worthwhile to move away from EVP_PKEY_CTX towards operation specific data types?  We have a lot of places where we would like to know what the key is going to be used for but don't.

This doesn't mean we cannot retain the EVP_PKEY_CTX underneath.

levitte commented:
> Just a thought, but might it be worthwhile to move away from EVP_PKEY_CTX towards operation specific data types? We have a lot of places where we would like to know what the key is going to be used for but don't.

So basically, add a more complete EVP_SIGNATURE API (along with a EVP_SIGNATURE_CTX), among others. I've had that in mind, and think I wrote an early proposal of that sort... 

> This doesn't mean we cannot retain the EVP_PKEY_CTX underneath.

"underneath"? Not sure why. For all intents and purposes, EVP_PKEY_CTX could be seen as an operation type plus the union of EVP_SIGNATURE_CTX and other operation specific types, should they spring into existence.

(type names may be discussed, I'm just modelling on the current trends)

paulidale commented:
> So basically, add a more complete EVP_SIGNATURE API (along with a EVP_SIGNATURE_CTX), among others. I've had that in mind, and think I wrote an early proposal of that sort...

In no way did I suggest more complex, quite the opposite.  The current EVP_PKEY_CTX APIs are far more complex and confusing than any reasonable alternative.  Honestly, look at our PKEY lifecycle diagram and weep:

[<img src="https://www.openssl.org/docs/manmaster/man7/img/pkey.png">](https://www.openssl.org/docs/manmaster/man7/life_cycle-pkey.html)

I doubt any operation specific APIs could be even remotely as complicated.  Also note that this _excludes_ quite a number of operations (e.g. freeing and finshing) unlike the other lifecycle diagrams because the graph became too complex to render sensibly.


> "underneath"? Not sure why. For all intents and purposes, EVP_PKEY_CTX could be seen as an operation type plus the union of EVP_SIGNATURE_CTX and other operation specific types, should they spring into existence.

All I'm suggesting here is the possibility that we introduce a new structure that wrappers the old:

```pascal
record EVP_SIGNATURE_CTX
   EVP_PKEY_CTX ctx
```

We gain type safety via the compiler without needing to rewrite massive amount of code.

The EVP_PKEY_CTX already includes the operation information, otherwise I'd suggest:

```pascal
record EVP_SIGNATURE_CTX
   const enum operation = evp_signature
   EVP_PKEY_CTX ctx
```

The latter might be better regardless, we could then deprecate the APIs that accept EVP_PKEY_CTX arguments and also deprecate the field in the PKEY_CTX structure.  There would likely also be other trimmings of the PKEY_CTX structure to be had.


Essentially I'm trying to minimising our code changes while adding type safety.

levitte commented:
Might I suggest that you make a separate discussion on that topic?

t8m commented:
The problem is that deprecating the existing EVP_PKEY_CTX based APIs is a huuuge breakage to existing users of these APIs which we've just recently (well... a few years ago) asked to move from the low-level APIs to these high level EVP_PKEY_CTX based ones. We would need to consciously decide on such change and IMO it should not be done in one corner of the EVP_PKEY_CTX apis alone. We should not introduce just EVP_SIGNATURE_CTX but we would need to introduce all the other cases from your above diagram as well. Besides if you strive for type safety - should there be a separate EVP_VERIFYSIG_CTX if there are separate inits for signature and verification??

I think this is opening a can of worms we probably do not want to open just now.

beldmit commented:
I'd also say we shouldn't deprecate EVP_PKEY_CTX API

t8m commented:
> I'd also say we shouldn't deprecate EVP_PKEY_CTX API

Yeah, and then adding EVP_SIGNATURE_CTX API would just add more complexity to the implementation and confusion to the users in what API should they choose.

romen commented:
@vdukhovni this is a relevant discussion https://github.com/openssl/openssl/pull/23240#issuecomment-1917532043

levitte commented:
There was a discussion on how ED25519 would fit in this API proposal, considering its double-pass nature (which I had frankly forgotten).

To me, that would make it perfectly suitable for oneshot functionality (which is essentially what `EVP_PKEY_sign()` is), and not at all suitable for a stream style functionality.  So, for this algorithm, it would be perfectly feasible to initiate it with this function (that's the function that was designed before the last fixup that replaced it with `EVP_PKEY_sign_primitive_init` and `EVP_PKEY_sign_data_init`):

```C
int EVP_PKEY_sign_init_ex2(EVP_PKEY_CTX *pctx,
                           EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
```

... and then call `EVP_PKEY_sign()` to perform the actual (oneshot) signature.

In provider terms, I would have an ED25519 implementation implement `OSSL_FUNC_SIGNATURE_SIGN_INIT` and `OSSL_FUNC_SIGNATURE_SIGN`, and that should be sufficient.  (As a matter of fact, this should conceptually be possible to do, and provide the use of the current EVP_PKEY_sign functionality, given an EVP_PKEY_CTX loaded with an ED25519 key).

t8m commented:
That's why you need all these APIs:

1. Oneshot primitive signature operation
2. Oneshot data signature operation
3. Streaming data signature operation

Let's look at which algos would implement what:
| Algo | Oneshot primitive | Oneshot data | Streaming data |
| -- | -- | -- | -- |
| RSA | x | - | - |
| sha256WithRSAEncryption | - | x | x |
| ECDSA | x | - | - |
| ecdsa-with-SHA256 | - | x | x |
| Ed25519 (pure) | - | x | - |
| Ed25519ph | x | x | x |

Let me explain the Ed25519ph - the primitive operation signs a hash value created with SHA-512, the oneshot data uses SHA-512 to hash arbitrary message (in one shot) and then signs the hash value with the Ed25519ph primitive and similarly for the streaming data signature.

As you can see for some algorithms all the three operations make sense and are doing something different.

levitte commented:
> Let me explain the Ed25519ph - the primitive operation signs a hash value created with SHA-512, the oneshot data uses SHA-512 to hash arbitrary message (in one shot) and then signs the hash value with the Ed25519ph primitive and similarly for the streaming data signature.

To me, the "primitive operation" sounds exactly like feeding a SHA-512 hash result to ED25519.  This is, incidently, exactly what RFC8032 says:

>    For Ed25519ph, phflag=1 and PH is SHA512 instead.  That is, the input
>    is hashed using SHA-512 before signing with Ed25519.

So in a way, one could say that it makes ED25519 the "primitive" for ED25519ph.

-----

I think we, again, need to actually define, at least for ourselves, what we want "primitive" to mean, on a conceptual level...  at least if we want to continue with that distinction.

levitte commented:
I a chat, @t8m and I came to the following definition of "primitive":

- an algorithm where the user usually must do some extra work with the input for safe or standard compliant usage

I'll try to work that into the design text

levitte commented:
There's been an exchange in #23240 that lead me to think that "primitive" isn't the best name for what's been discussed.  How about "digest" or "prehash":

``` C
int EVP_PKEY_sign_init_for_digest(EVP_PKEY_CTX *pctx,
                                  EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
```

I think that gives a better idea what this is supposed to do.

While I'm at it, "message" might also be a better name than "data"...

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

This modifies the design - yet again - as discussions have revealed aspects
that weren't immediately apparent.

t8m commented:
> (I imagine that `EVP_PKEY_sign_init_for_digest` call with a fetched "RSA" would refuse to work with `RSA_NO_PADDING`, but would otherwise work just as `EVP_PKEY_sign_init()` does with an RSA pkey)

Heh, and that's why I hate the `EVP_PKEY_sign_init_for_digest` name. I really do not understand what is the problem with the primitive word.

levitte commented:
> > (I imagine that `EVP_PKEY_sign_init_for_digest` call with a fetched "RSA" would refuse to work with `RSA_NO_PADDING`, but would otherwise work just as `EVP_PKEY_sign_init()` does with an RSA pkey)
> 
> Heh, and that's why I hate the `EVP_PKEY_sign_init_for_digest` name. I really do not understand what is the problem with the primitive word.

In light of PKCS#1, where the cryptographic primitives are RSAEP, RSADP, RSASP1 and RSAVP1, using "primitive" for something that's a bit higher level can be quite confusing.  Meanwhile, the *intent* with this form is explicitly to take a pre-hash as input, right?  I find it more clear if the function name reflects that intention.

> I have strong opinion the existing EVP_PKEY_sign_init() must be internally just an alias (apart from the EVP_SIGNATURE argument) for the primitive sign init (or whatever you call it). Otherwise this will be unimplementable properly on the provider side. Or, in other words, on the provider side you have to reflect the difference between the primitive and data inits otherwise the same confusion that is on the EVP API side will happen on the provider side.

Agreed, and if you look at the updated design, that's what I did.

levitte commented:
> I really do not understand what is the problem with the primitive word.

It's a source of confusion.  For example, if you come from reading PKCS#1, the primitives are cryptographic primitives are RSAEP, RSADP, RSASP1 and RSAVP1.  They take raw data (represented as numbers) with no further assumptions made on them.  That's a fairly intuitive use of the word "primitive", actually meaning *primitive*.

Of course, `EVP_PKEY_sign()` gives access to that sort of functionality under certain conditions (`RSA_NO_PADDING` for RSA keys, for example), but it also gives access to more functionality which, from a PKCS#1 perspective, is going away from being primitive.  *My intuition* would have me believe that something called `EVP_PKEY_sign_primitive()` gives me access to the `RSA_NO_PADDING` functionality for RSA keys, which is not at all what is intended.  So circling back to intention, using a name which indicates that the intended input is digests seems clearer.

levitte commented:
I kind of got a sort of "aha!" moment when I saw @davidben talk in terms of SignDigest and SignMessage, back in #23240.  That kinda connected the dots for me.

t8m commented:
The problem is that this function has to be able to do both - the sign primitive and the sign digest operations. But yeah, I am not going to insist on the `primitive` name.

vdukhovni commented:
> I kind of got a sort of "aha!" moment when I saw @davidben talk in terms of SignDigest and SignMessage, back in #23240. That kinda connected the dots for me.

I was hoping you'd see it that way. That is, instead of thinking in terms of how things are implemented, we should focus on the semantic properties of the APIs. And as @davidben pointed out, these are:

1. An IUF (init, update, final) API for digesting a *message* with a given signature algorithm, which may be a combination of a public key algorithm and choice of digest, or the digest may be a fixed feature of the signature algorithm as in Ed25519ph, where it is always SHA2-512. If the digest is configurable, the signature will typically cover not just the digest value, but also the choice of digest algorithm. Additional algorithm-specific parameters may be applicable.
2. A one-shot *message* signature, as in pure EdDSA (25519 or 448). Any internal digest is implicit, however there may still be additional parameters (e.g. the optional "context" in Ed25519ctx and Ed448ctx).
3. A raw one-shot signature of input that the caller should have pre-hashed, but this is an unsafe interface, mostly IMHO for testing debugging (as with "`rsautl -raw`, that I use from time to time to *peek* inside RSA signed blobs).

We're presently missing a IUF APIs for Ed25519ph and Ed448ph, which would be nice to also expose via `pkeyutl(1)`.

David

levitte commented:
What we *don't* have is a IUF API that uses a full, explicitly fetched EVP_SIGNATURE.  That would be useful for, like you said, ED25519ph, but also sha256WithRSAEncryption.  That's what this design is for.

vdukhovni commented:
> What we _don't_ have is a IUF API that uses a full, explicitly fetched EVP_SIGNATURE. That would be useful for, like you said, ED25519ph, but also sha256WithRSAEncryption. That's what this design is for.

Yes. Does the prior discussion then suggest a better name than "primitive"? Given IUF, we want some sort of context, that may as well capture at least the full signature algorithm (RSAwithXXX or Ed25519ph, ...), and perhaps also the key (unless that's part of "init", with "create" or "new" allocating the context).

Do you have a fully fleshed out IUF API (modulo its name)?

levitte commented:
This design is all about extending the `EVP_PKEY_sign` function set with I for explicitly fetched EVP_SIGNATURE and specific uses, along with the U and F functions.  The idea is that EVP_PKEY_CTX should be *preloaded* with the key to be used, and the provider interface gets an added query function to find out what key types a signature implementation supports.

This actually summarizes what's written in the design file.  I encourage you to have a closer look.

t8m commented:
Of course one option would be to call the new function for the _sign primitive_ or _sign digest_ operation just EVP_PKEY_sign_init_ex().

Whether that makes it good enough name to not be confused with the EVP_PKEY_sign_message_init() I do not know.

levitte commented:
> Of course one option would be to call the new function for the _sign primitive_ or _sign digest_ operation just EVP_PKEY_sign_init_ex().

That was my original design (it would have to be `_ex2`, `EVP_PKEY_sign_init_ex()`)...  so we're kind of coming full circle.

t8m commented:
> > Of course one option would be to call the new function for the _sign primitive_ or _sign digest_ operation just EVP_PKEY_sign_init_ex().
> 
> That was my original design (it would have to be `_ex2`, `EVP_PKEY_sign_init_ex()` already exists)... so we're kind of coming full circle.

Yeah, unfortunately, given there is no better name that would encompass both primitive (i.e. no padding in case of RSA) and less-primitive operations (i.e. PKCS1 1.5 or PSS padding).

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,189 @@
+#### Signing a stream
+
+``` C
+int EVP_PKEY_sign_init_for_message(EVP_PKEY_CTX *pctx,
````
I would call it EVP_PKEY_sign_message_init().

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,189 @@
+                         const unsigned char *in, size_t inlen);
+int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,
+                        unsigned char *sig, size_t *siglen);
+```
````
We will also need a oneshot. I.e., EVP_PKEY_sign_message(). And I would change those two above to EVP_PKEY_sign_message_update() and EVP_PKEY_sign_message_final().

Sashan commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,189 @@
+### Making things less confusing with distinct function names
+
+So far, `EVP_PKEY_sign()` and friends are only expected to act as
+"primitive" functions.  This design essentially proposes an extension to
```
I think we should not be using `primitve function` here. I think this paragraph  should try to untangle the confusion that comes from mixing up two distinct classes here:
   - API type which is a _one-shot_ vs. _streaming_ (IUF as @vdukhovni noted)
   - operation type which is either a cryptographic primitive itself (particular digest/encryption/sign/verify alg. such as AES or SHA256) or a compound operation which is made of more than one cryptographic primitive (such as digest + sign) 

also putting `primitive function` to google will get me to calculus which is even more confusing. So I would not be stealing terms from math here.

levitte replied:
I'm actually stealing "primitive" from [PKCS#1](https://www.rfc-editor.org/rfc/rfc8017#section-5) (`EVP_PKEY_sign` (etc) functions [actually, the provider] does give direct access to them, for RSA, under the condition that the flag `RSA_NO_PADDING` is set and that no hash function has been declared with `EVP_PKEY_CTX_set_signature_md()`, and with the [I2OSP and OSP2I](https://www.rfc-editor.org/rfc/rfc8017#section-4) primitives implied).

I'm wrong when I'm saying "only", though, since also message hashes are supported by our providers (and are properly encoded/decoded when the correct flags are set, and the signature hash is specified).

levitte replied:
I've expanded on this text to clarify the current functionality a bit more

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

This mostly expands the explanatory text, but also makes a statement on how
the new functionality is designed.

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

levitte commented:
@vdukhovni, you've asked me to provide some demo code for you to look at.  I'm currently working on the implementation in #23416, which includes enhanced documentation of both EVP_PKEY_sign.pod and EVP_PKEY_verify.pod, with a number of examples in each.  Is that good enough to at least eyeball?

(all examples refer to "RSA-SHA256", which I haven't finished implementing next, but am currently working on, with the hope to be done today or tomorrow)

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

levitte commented:
After OTC discussions today, I went along with @t8m's proposal for the application API.  I redressed and simplified the provider interface accordingly.

levitte commented:
> I redressed and simplified the provider interface accordingly.

Never mind, found bugs in my thinking.  Next update will fix them

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -34,98 +34,90 @@ So far, `EVP_PKEY_sign()` and friends are only expected to act on the
-                         const unsigned char *in, size_t inlen);
-int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,
-                        unsigned char *sig, size_t *siglen);
+int EVP_PKEY_verify_ex2(EVP_PKEY_CTX *pctx,
```
I think you meant `EVP_PKEY_verify_init_ex2` here?

levitte replied:
Oops.  Yes

t8m commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
````diff
@@ -0,0 +1,200 @@
+### For signing a stream with `EVP_SIGNATURE`
+
+``` C
+int EVP_PKEY_sign_message_init(EVP_PKEY_CTX *pctx,
+                               EVP_SIGNATURE *algo,
+                               const OSSL_PARAM params[]);
+int EVP_PKEY_sign_message_update(EVP_PKEY_CTX *ctx,
+                                 const unsigned char *in,
+                                 size_t inlen);
+int EVP_PKEY_sign_message_final(EVP_PKEY_CTX *ctx,
+                                unsigned char *sig,
+                                size_t *siglen);
````
What about the one-shot `EVP_PKEY_sign_message()` function? Please note that for example ED25519 requires two passes over the message -> this has to be a one-shot operation. Yeah, you could say that you can call EVP_PKEY_sign_message_update only once for ED25519 but that sounds like a kludge.

levitte replied:
EVP_PKEY_sign() can be used just as well.  You did propose an alias macro, something I'm mentioning higher up.  Do you actually want me to spell that out here?

t8m replied:
Yeah, please add it for clarity.

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

Special one for those who really want the oneshot aliases

slontis commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+                                 EVP_SIGNATURE *algo,
+                                 const OSSL_PARAM params[]);
+/* Signature setter */
+int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
```
Is this name ok?
It doesnt match all the others here (the CTX bit) - should we keep it consistent with the others..

Why doesnt the name have _verify_ in its name if it is only used for verification?

Having a seperate function is good for the case where the signature is needed early. For the other case it causes an unnecessary mem copy since the signature cant be applied via the message_final().
Is this ok?

t8m replied:
What about calling the function EVP_PKEY_CTX_set0_signature() or EVP_PKEY_verify_message_set0_signature()?

And use octet ref instead of octet string param. Of course that means usage in applications for the case where the message needs to be set before calling the update function(s) will be more complicated. Maybe the extra allocation/memcpy()/free call is not really relevant given the complexity of the signature verification.

levitte replied:
> Is the EVP_PKEY_CTX_set_signature name ok?
> It doesnt match all the others here (the CTX bit) - should we keep it consistent with the others, even if it is wrong.

I modelled that after already existing patterns, see EVP_PKEY_CTX_set_signature_md() and EVP_PKEY_CTX_set_rsa_padding().

> Why doesnt the name have _verify_ in its name if it is only used for verification?

Because the *thing* that's being set is the signature to verify against.  We don't have a *thing* called "verification", do we?

> Having a separate function is good for the exceptional case where the signature is needed early. For the normal case it causes an unnecessary memcopy since the signature cant be applied via the message_final(). Is this ok?

This was a choice to avoid causing confusion by allowing two ways to set that signature.  What happens if someone gives the signature to verify against using both EVP_PKEY_CTX_set_signature() and EVP_PKEY_verify_message_final() (if we imagine the latter with the added arguments to set the signature that way)?

levitte replied:
> What about calling the function EVP_PKEY_CTX_set0_signature() or EVP_PKEY_verify_message_set0_signature()?

I could go for either.

> And use octet ref instead of octet string param.

Which means making it another provider function...  well, considering that a signature is mandatory for verification, that makes sense.

slontis replied:
I prefer EVP_PKEY_verify_message_set0_signature() since it matches the other (wrong) ones :)

levitte replied:
set0?  set1?

levitte commented:
Considering that the implementation [#23416] is now merged, perhaps this should be merged as well?  For the signature setting function, @slontis and I concluded that enhancements can be done later, i.e. don't have to be resolved nownownow.  See this review: https://github.com/openssl/openssl/pull/23416#pullrequestreview-2249110642

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+Public API - Add variants of `EVP_PKEY_CTX` functionality
+---------------------------------------------------------
+
+Through OTC discussions, it's been determined that the most suitable APIs to
+touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
+`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
+associated functions.  They can be extended to accept an explicitly fetched
+algorithm of the right type, and to be able to process infinite amounts of
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
```
```suggestion
Through OTC discussions, it's been determined that the most suitable APIs to
touch are the `EVP_PKEY_` functions.
Specifically, `EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()`
and related functions.
They can be extended to accept an explicitly fetched algorithm of the right type,
and to be able to incrementally process indefinite length data streams when the
fetched algorithm permits it (typically, algorithms like ED25519 or RSA-SHA256).
```
Does Ed25519 really support streaming? Or are you thinking of Ed25519ph?
It was my impression that pure Ed25519 required the entire message in memory
and did not support IUF-style (Init, Update, Final) signing.

levitte replied:
Ah, I should remove ED25519 from there, 'cause indeed, the entire length must be known from the start, which had escaped me when I wrote this passage.

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
+
+It must be made clear that the added functionality can not be used to
+compose an algorithm from different parts.  For example, it's not possible
+to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
+specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
+`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
+other input that's limited to the modulus size of the RSA pkey.
```
Editorial nit aside, I find this paragraph confusing. Please explain what you mean more clearly.
```suggestion
It must be made clear that the added functionality cannot be used to
compose an algorithm from different parts.  For example, it's not possible
to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
other input that's limited to the modulus size of the RSA pkey.
```

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+
+### Making things less confusing with distinct function names
+
+So far, `EVP_PKEY_sign()` and friends are only expected to act on the
```
```suggestion
Till now, `EVP_PKEY_sign()` and friends are only expected to act on the
```

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+messages to be passed, in a streaming style through an *update* and a
+*final* function.
+
+Discussions have revealed that it is potentially confusing to confound the
```
```suggestion
Discussions have revealed that it is potentially confusing to conflate the
```

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+current functionality with streaming style functionality into the same name,
+so this design separates those out with specific init / update / final
+functions for that purpose.  For oneshot functionality, `EVP_PKEY_sign()`
+and `EVP_PKEY_verify()` remain supported, through an alias for the
+application.
```
```suggestion
and `EVP_PKEY_verify()` remain supported.
```
If you feel that the mechanism by which they remain supported should be explained, the explanation should be less terse. It is not clear what "an alias for the application" meant.

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+
+### Making it possible to verify with an early signature
+
+There are newer verifying algorithms that need to receive the signature
+before processing the data.  This is particularly important for streaming
+functionality.  This design proposes a mechanism to accomodate this for the
+streaming style functionality.
```
```suggestion
Some more recent verification algorithms need to obtain the signature
before processing the data.
This is particularly important for streaming modes of operation.
This design proposes a mechanism to accomodate these algorithms
and modes of operation.
```

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+int EVP_PKEY_verify_message_final(EVP_PKEY_CTX *ctx);
+
+#define EVP_PKEY_verify_message(ctx,sig,siglen,tbs,tbslen) \
+    EVP_PKEY_sign(ctx,sig,siglen,tbs,tbslen)
```
This seems to be a cut/paste error from the signature side.

levitte replied:
Yes it was

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+
+### For verify_recover with `EVP_SIGNATURE`
+
+According to feedback, a stream style interface is uninteresting for
+verify_recover, so we only specify a new init function.
```
```suggestion
Preliminary feedback suggests that a streaming interface is uninteresting for
verify_recover, so we only specify a new init function.
```

vdukhovni commented on doc/designs/functions-for-explicitly-fetched-signature-algorithms.md:
```diff
@@ -0,0 +1,205 @@
+Requirements on the providers
+-----------------------------
+
+Because it's not immediately obvious from a composite algorithm name what
+key type it requires / supports, at least in code, allowing the use of an
```
```suggestion
Because it isn't always obvious from a composite algorithm name what
key type it requires / supports, allowing the use of an
```
When you say "key type", what do you mean?
- Public vs. Private?
- Particular asymmetric algorithm?
- Distinctions like RSA unrestricted vs. RSA PSS?
- Something else?

levitte replied:
We use the term "key type" in other places for "RSA", "DSA", "EC"...

I a provider context, these are essentially keymgmt names.

vdukhovni replied:
> We use the term "key type" in other places for "RSA", "DSA", "EC"...
> 
> I a provider context, these are essentially keymgmt names.

So, essentially the value of EVP_PKEY_type()? An in-context example could help avoid confusion in the minds of the uninitiated, perhaps: key type ("RSA", "EC", ...).
EVP_PKEY_type()

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

levitte committed:
fixup! [design] Functions for explicitly fetched signature algorithms

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 216 days ago

paulidale commented:
This possibly warrants back porting to relevant older branches.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been  updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

t8m commented:
Merged to the master, 3.4 and 3.5 branches. Thank you.
